client/src/pages/Chat.tsx

```diff
@@ -69,12 +66,54 @@ export const Chat = () => {
   }, [chat, user, navigate]);
 
   useEffect(() => {
-    // scroll to bottom of the chat when getting a new msg
+    // check if the scrolling element is in viewport
+    function isInViewport(element: HTMLElement) {
+      const rect = element.getBoundingClientRect();
+      return (
+        rect.top >= 0 &&
+        (rect.bottom - 50) <= (window.innerHeight || document.documentElement.clientHeight)
+      );
+    }
     if (scrollingRef.current) {
-      (scrollingRef.current as HTMLElement).scrollIntoView({ behavior: 'smooth' });
+      if (isInViewport(scrollingRef.current)) {
+        scrollToBottom('smooth');
+        setMsgDiff(0);
+      } else {
+        // if now in viewport - add button to scroll down
+        setMsgLength(prev => {
+          const diff = (messages?.length || 0) - prev;
+          setMsgDiff(prev => prev + diff);
+          return messages?.length || 0;
+        })
+      }
     }
   }, [messages]);
 
+  useEffect(() => {
+    // immediately scroll down when entering the page
+    if (scrollingRef.current) {
+      scrollToBottom('instant');
+    }
+  }, [scrollingRef.current]);
+
+  const handleScroll = (e: React.UIEvent<HTMLElement>) => {
+    const element = e.currentTarget;
+    if (element.scrollHeight - element.scrollTop - 1 <= element.clientHeight) {
+      setMsgDiff(0);
+    }
+  };
+
+  const scrollToBottom = (behavior: 'smooth' | 'instant' | 'auto') => {
+    (scrollingRef.current as HTMLElement).scrollIntoView({ behavior });
+    setMsgLength(messages.length);
+  }
+
+  const scrollDownClick = () => {
+    if (scrollingRef.current) {
+      scrollToBottom('smooth');
+    }
+  };
+
```

Comment:
Responsibility Separation:
It might be beneficial to separate the logic for handling the scroll event (handleScroll) from the scrolling behavior itself (scrollToBottom). This can make each function more focused on its specific responsibility.